Alternate short channel matching support

christiangnrd opened the pull request:
Closes #836 and closes #830.

If the channel is the prefix for one and only one installed channel, use that, otherwise old behaviour. This behaves more like the julia command line arguments

Should work with any installed channel including linked channels instead of a hard-coded few.

LilithHafner commented:
I don't like this because I could have Julia 1.6.2 as the only installed version and then run `julia +1` and get that version. Similarly, I could have Julia 1.10 installed and run `julia +1.1` and get version 1.10 instead.

The reasons this short matching works okay in Julia (and some other programs') command matching is that
- Their options are intelligently named to avoid misleading prefixing
- Their options are words or phrases that have a more diverse prefix set than version numbers

Whereas in this use case,
- Juliaup's set of version options varies depending on the installation (if we are filtering to installed channels)
- Juliaup's option names are already very concise so the benefit is less.

christiangnrd commented:
I've updated to not match on version numbers. The approach feels hacky but also I don't know if it needs t be more complicated. 

Unfortunately the new tests end up downloading 2 extra Julia versions so that might take up a lot more time. Maybe I should test with channels linked to already downloaded versions?

LilithHafner commented on src/bin/julialauncher.rs:
```diff
@@ -166,32 +166,49 @@ fn get_julia_path_from_channel(
+        .installed_channels
+        .keys()
+        .filter(|&item| {
+            item.starts_with(channel) && !item.starts_with("0") && !item.starts_with("1")
```
This is more general:
```suggestion
            item.starts_with(channel) && !item.chars().next().map(char::is_numeric).unwrap_or(false)
```

LilithHafner commented on src/bin/julialauncher.rs:
```diff
@@ -166,32 +166,49 @@ fn get_julia_path_from_channel(
     juliaupconfig_path: &Path,
     juliaup_channel_source: JuliaupChannelSource,
 ) -> Result<(PathBuf, Vec<String>)> {
+    let potential_matches: Vec<_> = config_data
+        .installed_channels
+        .keys()
+        .filter(|&item| {
+            item.starts_with(channel) && !item.starts_with("0") && !item.starts_with("1")
+        })
+        .cloned()
+        .collect();
+    let actual_channel = if potential_matches.len() == 1 {
+        potential_matches
+            .first()
+            .expect("length should be 1")
+            .as_str()
+    } else {
+        channel
+    };
```
A more efficient solution (though probably negligibly so)

```suggestion
    let actual_channel = if channel
        .chars()
        .next()
        .map(char::is_numeric)
        .unwrap_or(false)
    {
        channel
    } else {
        match only(
            config_data
                .installed_channels
                .keys()
                .filter(|&item| item.starts_with(channel)),
        ) {
            Some(val) => val,
            None => channel,
        }
    };
```

With a helper function

```rs
fn only<T, I>(mut iter: I) -> Option<T>
where
    I: Iterator<Item = T>,
{
    let first = iter.next();
    if iter.next().is_some() {
        // There are multiple elements
        return None;
    }
    first
}
```

Mostly this is a stylistic choice.

LilithHafner commented on src/bin/julialauncher.rs:
```diff
@@ -166,32 +166,49 @@ fn get_julia_path_from_channel(
     let channel_info = config_data
             .installed_channels
-            .get(channel)
+            .get(actual_channel)
```
`channel`/`actual_channel` is not a good pair of variable names, imo.

`requested_channel`/`expanded_channel` could be a better choice.

LilithHafner commented on tests/channel_selection.rs:
```diff
@@ -125,4 +125,63 @@ fn channel_selection() {
+    // At this point, installed channels are: 1.6.7, 1.7.3, 1.8.5
+
+    // Test that incomplete number matching does not autocomplete:
+    // https://github.com/JuliaLang/juliaup/pull/838#issuecomment-2206640506)
```
```suggestion
    // https://github.com/JuliaLang/juliaup/pull/838#issuecomment-2206640506
```

LilithHafner commented on tests/channel_selection.rs:
```diff
@@ -125,4 +125,63 @@ fn channel_selection() {
+        .arg("print(VERSION)")
+        .env("JULIA_DEPOT_PATH", depot_dir.path())
+        .env("JULIAUP_DEPOT_PATH", depot_dir.path())
+        .env("JULIAUP_CHANNEL", "1.7.3")
```
```suggestion
```
I don't think we should set this env variable.

LilithHafner commented on tests/channel_selection.rs:
```diff
@@ -125,4 +125,63 @@ fn channel_selection() {
+    Command::cargo_bin("julia")
+        .unwrap()
+        .arg("+1.8")
+        .arg("-e")
+        .arg("print(VERSION)")
```
These are also unnecessary

christiangnrd replied:
This might be a stupid question but how does the test not get stuck in the REPL without the "-e"?

LilithHafner replied:
Good point; that is necessary (or -v or something)

christiangnrd replied:
I was asking because the tests pass without them and I was genuinely wondering what made it work. I'll add -v to be safe because though.

LilithHafner reviewed (commented):
This works; I left a few comments, but fundamentally it's a design choice between this and #836. Do we want `+r` to always be `release`, or to be either `rc` or `release` depending on which channels the use has installed. I lean toward the simplicity of #836.

christiangnrd commented:
I've addressed your feedback and implemented all your suggestions.

Regarding this PR or #836, I wrote this because I sometimes write more than just the first letter, and I want this feature to work with linked channels. Also, for someone like me who sometimes tries out PRs but rarely more than one at a time, once #903 is a thing, it'll be nice to be able to call `julia +pr` instead of having to remember the pull request number.

LilithHafner commented:
@rakeshksr @christiangnrd and espeically @davidanthoff, could we discuss which option we prefer (#838 vs #836) both are imo ready to merge but we need to pick which.

christiangnrd commented:
My justification for preferring this approach are in my previous comment. To add to that, I always have a `dev` channel linked to my local branch of Julia, I feel like it would be quite unintuitive to have `+b` work for beta, and `+l` work for lts, but `+d` error out for dev.

davidanthoff commented:
I think I've finally am getting clarity what I'm thinking about this, see https://github.com/JuliaLang/juliaup/issues/830.
